src/coprocessor/statistics/analyze.rs

```diff
@@ -271,6 +291,232 @@ impl<S: Snapshot> RequestHandler for AnalyzeContext<S> {
+    fm_sketches: Vec<FmSketch>,
+    rng: StdRng,
+    total_sizes: Vec<i64>,
+    row_buf: Vec<u8>,
+}
+
+impl Default for RowSampleCollector {
+    fn default() -> Self {
+        RowSampleCollector {
+            samples: BinaryHeap::new(),
+            null_count: vec![],
+            count: 0,
+            max_sample_size: 0,
+            fm_sketches: vec![],
+            rng: StdRng::from_entropy(),
+            total_sizes: vec![],
+            row_buf: Vec::new(),
+        }
+    }
+}
+
+impl RowSampleCollector {
+    fn new(
+        max_sample_size: usize,
+        max_fm_sketch_size: usize,
+        col_and_group_len: usize,
+    ) -> RowSampleCollector {
+        RowSampleCollector {
+            samples: BinaryHeap::new(),
+            null_count: vec![0; col_and_group_len],
+            count: 0,
+            max_sample_size,
+            fm_sketches: vec![FmSketch::new(max_fm_sketch_size); col_and_group_len],
+            rng: StdRng::from_entropy(),
+            total_sizes: vec![0; col_and_group_len],
+            row_buf: Vec::new(),
+        }
+    }
+
+    pub fn collect_column_group(
+        &mut self,
+        columns_val: &[Vec<u8>],
+        column_groups: &[tipb::AnalyzeColumnGroup],
+    ) {
+        let col_len = columns_val.len();
+        for i in 0..column_groups.len() {
+            self.row_buf.clear();
+            let offsets = column_groups[i].get_column_offsets();
+            let mut has_null = true;
+            for j in 0..offsets.len() {
+                if columns_val[j][0] != NIL_FLAG {
+                    has_null = false
+                }
+                self.total_sizes[col_len + i] += columns_val[i].len() as i64
+            }
+            // We only maintain the null count for single column case.
+            if has_null && offsets.len() == 1 {
+                self.null_count[col_len + i] += 1;
+                continue;
+            }
+            // Use a in place murmur3 to replace this memory copy.
+            for j in 0..offsets.len() {
+                self.row_buf.extend_from_slice(&columns_val[j]);
+            }
+            self.fm_sketches[col_len + i].insert(&self.row_buf);
+        }
+    }
+
+    pub fn collect_column(&mut self, columns_val: Vec<Vec<u8>>) {
+        for i in 0..columns_val.len() {
+            if columns_val[i][0] == NIL_FLAG {
+                self.null_count[i] += 1;
+                continue;
+            }
+            self.fm_sketches[i].insert(&columns_val[i]);
+            self.total_sizes[i] += columns_val[i].len() as i64;
+        }
+        self.sampling(columns_val);
+    }
+
+    pub fn sampling(&mut self, data: Vec<Vec<u8>>) {
+        let cur_rng = self.rng.gen_range(0, i64::MAX);
+        if self.samples.len() < self.max_sample_size {
+            self.samples.push(Reverse((cur_rng, data)));
+            return;
+        }
+        if self.samples.len() == self.max_sample_size && self.samples.peek().unwrap().0.0 < cur_rng
```

Comment:
The `samples` is a max value heap? Here should be a min value heap?

Reply:
Rust std's heap is by default max-value heap. As we wrap the original type with `Reverse`, it is now a min heap.

Reply:
If we keep the max value heap. All the value which is less than the max value in heap will never be replaced?

Reply:
OK. Got it.

Reply:
* Our goal is to get maximum k elements.
* The heap stores maximum k elements of first n elements, and as the heap is a min-heap, the minimum one will be at the top.
* If the next element is larger than minimum of current maximum k elements, then pop current top and add new one.